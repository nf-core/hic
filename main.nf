#!/usr/bin/env nextflow
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    nf-core/hic
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Github : https://github.com/nf-core/hic
    Website: https://nf-co.re/hic
    Slack  : https://nfcore.slack.com/channels/hic
----------------------------------------------------------------------------------------
*/

nextflow.enable.dsl = 2

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT FUNCTIONS / MODULES / SUBWORKFLOWS / WORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

include { HIC  } from './workflows/hic'
include { PIPELINE_INITIALISATION } from './subworkflows/local/utils_nfcore_hic_pipeline'
include { PIPELINE_COMPLETION     } from './subworkflows/local/utils_nfcore_hic_pipeline'

include { getGenomeAttribute      } from './subworkflows/local/utils_nfcore_hic_pipeline'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    GENOME PARAMETER VALUES
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

params.fasta = getGenomeAttribute('fasta')
params.bwt2_index = getGenomeAttribute('bowtie2')
params.bwa_index = getGenomeAttribute('bwamem2')

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    NAMED WORKFLOWS FOR PIPELINE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//
// WORKFLOW: Run main analysis pipeline depending on type of input
//
workflow NFCORE_HIC {

    take:
    samplesheet // channel: samplesheet read in from --input

    main:

    //
    // WORKFLOW: Run pipeline
    //
    HIC (
        samplesheet
    )

    emit:
    multiqc_report = HIC.out.multiqc_report // channel: /path/to/multiqc_report.html

}
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow {

    main:

    //
    // SUBWORKFLOW: Run initialisation tasks
    //
    PIPELINE_INITIALISATION (
        params.version,
        params.help,
        params.validate_params,
        params.monochrome_logs,
        args,
        params.outdir,
        params.input
    )

    //
    // WORKFLOW: Run main workflow
    //
    NFCORE_HIC (
        PIPELINE_INITIALISATION.out.samplesheet
    )

    //
    // SUBWORKFLOW: Run completion tasks
    //
    PIPELINE_COMPLETION (
        params.email,
        params.email_on_fail,
        params.plaintext_email,
        params.outdir,
        params.monochrome_logs,
        params.hook_url,
        NFCORE_HIC.out.multiqc_report
    )
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
