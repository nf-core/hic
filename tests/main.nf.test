nextflow_pipeline {

    name "Test pipeline"
    script "../main.nf"
    tag "PIPELINE"
    config "conf/test.config"

    test("Run with profile test") {

        when {
            params {
                outdir = "$outputDir"
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(UTILS.removeNextflowVersion("$outputDir/pipeline_info/nf_core_pipeline_software_mqc_versions.yml")).match("software_versions") },
                { assert snapshot(
                    workflow.trace.tasks().size(),
                    // FIXME Not deterministic on CI but works locally
                    // path("$outputDir/compartments").list(),
                    path("$outputDir/distance_decay").list(),
                    path("$outputDir/hicpro/stats/SRR4292758/SRR4292758.mpairstat"),
                    path("$outputDir/hicpro/stats/SRR4292758/SRR4292758_allValidPairs.mergestat"),
                    path("$outputDir/hicpro/valid_pairs").list(),
                    // path("$outputDir/multiqc/multiqc_data").list(),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_boundaries.bed"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_boundaries.gff"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_domains.bed"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_score.bedgraph"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_tad_score.bm"),
                    path("$outputDir/tads/insulation").list(),
                ).match()},

                { assert new File("$outputDir/multiqc/multiqc_report.html").exists() },
                { assert new File("$outputDir/compartments/SRR4292758.2000_compartments.cis.bw").exists() },
                { assert new File("$outputDir/compartments/SRR4292758.2000_compartments.cis.lam.txt").exists() },
                { assert new File("$outputDir/compartments/SRR4292758.2000_compartments.cis.vecs.tsv").exists() },
                { assert new File("$outputDir/fastqc/SRR4292758_1_fastqc.html").exists() },
                { assert new File("$outputDir/fastqc/SRR4292758_2_fastqc.html").exists() },
                { assert new File("$outputDir/hicpro/mapping/SRR4292758_0_bwt2pairs.bam").exists() },
                { assert new File("$outputDir/hicpro/stats/SRR4292758/SRR4292758.mRSstat").exists() },
                { assert new File("$outputDir/hicpro/stats/SRR4292758/SRR4292758.null.mmapstat").exists() },
                { assert new File("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_zscore_matrix.cool").exists() },
            )
        }
    }

    test("Should split fastqs") {

        when {
            params {
                outdir = "$outputDir"
                split_fastq = true
                fastq_chunks_size = 125000
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(UTILS.removeNextflowVersion("$outputDir/pipeline_info/nf_core_pipeline_software_mqc_versions.yml")).match("split_software_versions") },
                { assert snapshot(
                    workflow.trace.tasks().size(),
                    // FIXME Not deterministic on CI but works locally
                    // path("$outputDir/compartments").list(),
                    path("$outputDir/distance_decay").list(),
                    path("$outputDir/hicpro/stats/SRR4292758/SRR4292758_allValidPairs.mergestat"),
                    // FIXME path("$outputDir/hicpro/valid_pairs").list(),
                    // path("$outputDir/multiqc/multiqc_data").list(),
                    // path("$outputDir/multiqc/multiqc_data").list(),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_boundaries.bed"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_boundaries.gff"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_domains.bed"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_score.bedgraph"),
                    path("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_tad_score.bm"),
                    path("$outputDir/tads/insulation").list(),
                ).match()},

                { assert new File("$outputDir/multiqc/multiqc_report.html").exists() },
                { assert new File("$outputDir/compartments/SRR4292758.2000_compartments.cis.bw").exists() },
                { assert new File("$outputDir/compartments/SRR4292758.2000_compartments.cis.lam.txt").exists() },
                { assert new File("$outputDir/compartments/SRR4292758.2000_compartments.cis.vecs.tsv").exists() },
                { assert new File("$outputDir/fastqc/SRR4292758_1_fastqc.html").exists() },
                { assert new File("$outputDir/fastqc/SRR4292758_2_fastqc.html").exists() },
                { assert new File("$outputDir/hicpro/mapping/SRR4292758_0_bwt2pairs.bam").exists() },
                { assert new File("$outputDir/hicpro/mapping/SRR4292758_1_bwt2pairs.bam").exists() },
                { assert new File("$outputDir/hicpro/mapping/SRR4292758_2_bwt2pairs.bam").exists() },
                { assert new File("$outputDir/hicpro/mapping/SRR4292758_3_bwt2pairs.bam").exists() },
                { assert new File("$outputDir/hicpro/stats/SRR4292758/SRR4292758.mpairstat").exists() },
                { assert new File("$outputDir/hicpro/stats/SRR4292758/SRR4292758.mRSstat").exists() },
                { assert new File("$outputDir/hicpro/stats/SRR4292758/SRR4292758.null.mmapstat").exists() },
                { assert new File("$outputDir/tads/hicExplorer/SRR4292758.1000_balanced_hicfindtads_zscore_matrix.cool").exists() },
            )
        }
    }
}
