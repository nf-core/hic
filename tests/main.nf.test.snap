{
    "split_software_versions": {
        "content": [
            "{BOWTIE2_ALIGN={bowtie2=2.5.2, samtools=1.18, pigz=2.6}, BOWTIE2_ALIGN_TRIMMED={bowtie2=2.5.2, samtools=1.18, pigz=2.6}, BOWTIE2_BUILD={bowtie2=2.5.2}, COMBINE_MATES={python=3.9.12}, COOLER_BALANCE={cooler=0.9.2}, COOLER_CLOAD={cooler=0.9.2}, COOLER_DUMP={cooler=0.9.2}, COOLER_MAKEBINS={cooler=0.9.2}, COOLER_ZOOMIFY={cooler=0.9.2}, COOLTOOLS_EIGSCIS={cooltools=0.5.1}, COOLTOOLS_INSULATION={cooltools=0.5.1}, CUSTOM_GETCHROMSIZES={getchromsizes=1.16.1}, FASTQC={fastqc=0.12.1}, GET_RESTRICTION_FRAGMENTS={python=3.9.12}, GET_VALID_INTERACTION={python=3.9.12}, HICPRO2PAIRS={pairix=0.3.7}, HIC_FIND_TADS={hicexplorer=3.7.2}, HIC_PLOT_DIST_VS_COUNTS={hicexplorer=3.7.2}, MERGE_BOWTIE2={samtools=1.15.1}, MERGE_STATS={python=3.9.12}, MERGE_VALID_INTERACTION={sort=8.3}, SPLIT_COOLER_DUMP={cooler=null}, TRIM_READS={gzip=1.1}, Workflow={nf-core/hic=v2.2.0dev}}"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-09T18:39:34.552071308"
    },
    "software_versions": {
        "content": [
            "{BOWTIE2_ALIGN={bowtie2=2.5.2, samtools=1.18, pigz=2.6}, BOWTIE2_ALIGN_TRIMMED={bowtie2=2.5.2, samtools=1.18, pigz=2.6}, BOWTIE2_BUILD={bowtie2=2.5.2}, COMBINE_MATES={python=3.9.12}, COOLER_BALANCE={cooler=0.9.2}, COOLER_CLOAD={cooler=0.9.2}, COOLER_DUMP={cooler=0.9.2}, COOLER_MAKEBINS={cooler=0.9.2}, COOLER_ZOOMIFY={cooler=0.9.2}, COOLTOOLS_EIGSCIS={cooltools=0.5.1}, COOLTOOLS_INSULATION={cooltools=0.5.1}, CUSTOM_GETCHROMSIZES={getchromsizes=1.16.1}, FASTQC={fastqc=0.12.1}, GET_RESTRICTION_FRAGMENTS={python=3.9.12}, GET_VALID_INTERACTION={python=3.9.12}, HICPRO2PAIRS={pairix=0.3.7}, HIC_FIND_TADS={hicexplorer=3.7.2}, HIC_PLOT_DIST_VS_COUNTS={hicexplorer=3.7.2}, MERGE_BOWTIE2={samtools=1.15.1}, MERGE_STATS={python=3.9.12}, MERGE_VALID_INTERACTION={sort=8.3}, SPLIT_COOLER_DUMP={cooler=null}, TRIM_READS={gzip=1.1}, Workflow={nf-core/hic=v2.2.0dev}}"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-09T18:36:56.07112172"
    },
    "Run with profile test": {
        "content": [
            36,
            [
                "SRR4292758_distcount.png:md5,5314c9ad16a1e857682fde4e584103bc",
                "SRR4292758_distcount.txt:md5,bac196d56bcb489c2ef6fe2ea26e07c9"
            ],
            "SRR4292758.mpairstat:md5,281aa85283e57308187c33a5bad7dc82",
            "SRR4292758_allValidPairs.mergestat:md5,b0a81a4d69ea8e44495a714ac87701e8",
            [
                "SRR4292758.allValidPairs:md5,55e00a772766779f5d5c2bf2b6fcef6d",
                [
                    "SRR4292758_contacts.pairs.gz:md5,201d59212cf443fe5da09dbddeef8677",
                    "SRR4292758_contacts.pairs.gz.px2:md5,6275247d930753121aa026870864aa25"
                ]
            ],
            "SRR4292758.1000_balanced_hicfindtads_boundaries.bed:md5,d41d8cd98f00b204e9800998ecf8427e",
            "SRR4292758.1000_balanced_hicfindtads_boundaries.gff:md5,d41d8cd98f00b204e9800998ecf8427e",
            "SRR4292758.1000_balanced_hicfindtads_domains.bed:md5,d41d8cd98f00b204e9800998ecf8427e",
            "SRR4292758.1000_balanced_hicfindtads_score.bedgraph:md5,da38c91d9b23820e290d2c987f27ab70",
            "SRR4292758.1000_balanced_hicfindtads_tad_score.bm:md5,81b33e5cd3ff14efac94c2065ac68fe1",
            [
                "SRR4292758.1000_balanced_insulation.tsv:md5,fd0a37e94600e9da4c8de0c58f07e3ed"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-09T18:36:56.085219308"
    },
    "Should split fastqs": {
        "content": [
            69,
            [
                "SRR4292758_distcount.png:md5,5314c9ad16a1e857682fde4e584103bc",
                "SRR4292758_distcount.txt:md5,bac196d56bcb489c2ef6fe2ea26e07c9"
            ],
            "SRR4292758_allValidPairs.mergestat:md5,b0a81a4d69ea8e44495a714ac87701e8",
            "SRR4292758.1000_balanced_hicfindtads_boundaries.bed:md5,d41d8cd98f00b204e9800998ecf8427e",
            "SRR4292758.1000_balanced_hicfindtads_boundaries.gff:md5,d41d8cd98f00b204e9800998ecf8427e",
            "SRR4292758.1000_balanced_hicfindtads_domains.bed:md5,d41d8cd98f00b204e9800998ecf8427e",
            "SRR4292758.1000_balanced_hicfindtads_score.bedgraph:md5,da38c91d9b23820e290d2c987f27ab70",
            "SRR4292758.1000_balanced_hicfindtads_tad_score.bm:md5,81b33e5cd3ff14efac94c2065ac68fe1",
            [
                "SRR4292758.1000_balanced_insulation.tsv:md5,fd0a37e94600e9da4c8de0c58f07e3ed"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-09T18:39:34.716731734"
    }
}